torch/_export/serde/serialize.py

```diff
@@ -194,7 +207,47 @@ def deserialize_metadata(metadata) -> Dict[str, str]:
     module_fqn = metadata.get("module_fqn")
     if module_fqn is not None:
         ret["module_fqn"] = module_fqn
-    # TODO(angelayi) add nn_module_stack and source_fn
+
+    def deserialize_meta_func(serialized_target: str):
+        module = None
+        if serialized_target.startswith("torch.nn"):
+            module = torch.nn
+            serialized_target_names = serialized_target.split(".")[2:]
+        elif serialized_target.startswith("torch"):
+            module = torch
+            serialized_target_names = serialized_target.split(".")[1:]
+        else:
+            return deserialize_operator(serialized_target)
+
+        target = module
+        for name in serialized_target_names:
+            if not hasattr(target, name):
+                return serialized_target
+            else:
+                target = getattr(target, name)
+        return target
+
+    if nn_module_stack_str := metadata.get("nn_module_stack"):
+        # Originally serialized to "fx_node_name:(orig_ref,type_str)"
+        nn_module_stack_list = nn_module_stack_str.split(";")
+        nn_module_stack = {}
+        for kv in nn_module_stack_list:
+            key_idx = kv.find(":")
+            key = kv[:key_idx]
+            assert kv[key_idx + 1] == "("
+            assert kv[-1] == ")"
+            values = kv[key_idx + 2: -1].split(",")
+            assert len(values) == 2
+            module = deserialize_meta_func(values[1])
+            nn_module_stack[key] = (values[0], module)
+        ret["nn_module_stack"] = nn_module_stack
+
+    if source_fn_str := metadata.get("source_fn"):
```

Comment:
who needs source_fn to be serialized?

Reply:
The source_fn is used to partition modules (torch.nn.linear) and functional modules (torch.nn.F.linear). It's almost like nn_module_stack but also works for functions that this node was decomposed from.

Reply:
Curious, is this just the fully qualified name of the function? Or something more structured that has been `str()`-ified?

Reply:
It's a tuple containing the fully qualified name of the node that was inserted before lowering to aten graph, and the function or module it was before decomposition. The function could also be a string name 😅